numpy/core/getlimits.py

```diff
@@ -208,23 +273,27 @@ def _register_known_types():
     # https://en.wikipedia.org/wiki/Quadruple-precision_floating-point_format#Double-double_arithmetic
     # These numbers have the same exponent range as float64, but extended number of
     # digits in the significand.
-    huge_dd = (umath.nextafter(ld(inf), ld(0))
-                if hasattr(umath, 'nextafter')  # Missing on some platforms?
-                else float64_ma.huge)
+    huge_dd = nextafter(ld(inf), ld(0), dtype=ld)
+    # As the smallest_normal in double double is so hard to calculate we set
+    # it to NaN.
+    smallest_normal_dd = NaN
+    # Leave the same value for the smallest subnormal as double
+    smallest_subnormal_dd = ld(nextafter(0., 1.))
```

Comment:
Calculating the value for smallest_normal in some architectures is messy, so after a lot of discussions and thoughts we defined that it is safer to leave this value as NaN because the previous value is wrong. Once this PR is merged, I'll open a new issue to discuss how we should approach the calculation of this value in a way where everyone feels confident with the resulting number.

Reply:
@wkschwartz I am not happy to have this NaN, but unless we get clarity for one more person agreeing on the correct value, I am unwilling to make this my decision. I _am_ willing to sign off someone else's research if I get a bit more background (or just a spec/prior art) – including an argument why the double-double specific storage scheme doesn't break e.g. your reasoning.

Even `macheps` is only an ["appropriate" value](https://www.ibm.com/docs/en/aix/7.1?topic=sepl-128-bit-long-double-floating-point-data-type), so we should be a bit careful...  I suspect that may give you the "most appropriate" value (i.e. the smallest number that has the full 106-bit of "guaranteed" mantissa, of which one is implicit).

That said, I also come to the a value of `np.exp2(-969.0)` which is identical to `np.log2(np.nextafter(0., 1.)) + 105` (105 and is just 52+53, since we include the implicit mantissa bit of the second float).  But, I have never worked with double-double and would like someone else to think it through first.

Reply:
I guess I'm confused about the target FP system here. Am I correct that this is for double-double implementations of [`long double`]? The [AIX 7.1 documentation] you linked to says explicitly that its 128-bit `long double` is _not_ compliant with C99 because it does not have a fixed precision (number of bits in the significand):

> The actual number of bits of precision can vary.... The ANSI C standard... requires that the number of base _b_ digits is fixed, which is not true for 128-bit long double numbers.

I didn't realize that this code was targeting variable-precision number formats when I wrote my first post. **I think I take back my suggestion above?**

You wrote

> if I get a bit more background (or just a spec/prior art)

I would argue that a variable-precision number format is not a floating-point format. Is it really appropriate to present information about such a format through `finfo` ("leaky abstraction", etc.)?† The definitions of _floating point system_ in Goldberg (1991), Higham (2002), and Muller, _et al._ (2010) require fixed precision. I don't know which theorems that numerical analysts rely on for designing accurate algorithms (Higham, 2002) apply to variable-precision formats. Muller, _et al._ (2010, Ch. 14) discusses them a little, offers some references, and says some not very nice things about them ("must be used with caution", etc.). I haven't read the whole chapter, so that's as much as I can offer at this point.

I guess my question is this: **What platforms that Numpy targets—aside from AIX's `xlc128` compiler—have a `long double` with variable precision?** It seems to me (but I immediately defer to your expertise here) that the rest of Numpy presupposes at least an FP format compliant with C99, no? **If AIX's `xlc128` compiler is the only such platform, then I think the data presented from `finfo` about `long double` should just directly target that one platform. In particular, I suppose that there _is no_ standard for double-double formats, they're always idiosyncratic where they exist, and they're not really even floating-point systems.**

For what it's worth, the [AIX 7.1 documentation] says

> When the value to be represented is in the denormal range, this representation provides no more precision than the 64-bit double-precision data type.

This suggests that `smallest_normal_dd` is the same as the smallest, positive normal in a `double`. (Also it should be the same as C's [`LDBL_MIN`](https://en.cppreference.com/w/c/types/limits#Limits_of_floating_point_types).)

### References

I'm copypasta-ing some references here from an paper I've been working on (FP-related, not double-double related). Maybe someone else can make use of some of these links to shed some light on this?

1. David Goldberg, “What Every Computer Scientist Should Know about Floating-Point Arithmetic” in _ACM Computing Surveys_, vol. 23, no. 1 (March 1991), 5–48. [doi:10.1145/103162.103163](https://doi.org/10.1145/103162.103163). *Available at* http://lux.dmcs.p.lodz.pl/ak/IEEE754_article.pdf.
2. Nicholas J. Higham. _Accuracy and Stability of Numerical Algorithms_, 2nd ed. Philadelphia: Society for Industrial and Applied Mathematics, 2002.
3. “IEEE Standard for Floating-Point Arithmetic” in _IEEE Std 754-2008_, 29 Aug. 2008, [doi:10.1109/IEEESTD.2008.4610935](https://doi.org/10.1109/IEEESTD.2008.4610935). _Available at_ https://irem.univ-reunion.fr/IMG/pdf/ieee-754-2008.pdf.
4. ISO/IEC 9899:1999, committee draft WG14 N1256, _Programming languages – C_, 2007-09-07. International Organization for Standardization. Geneva, Switzerland. _Available at_ http://www.open-std.org/jtc1/sc22/wg14/www/docs/n1256.pdf. This is not the published international standard, but the latest publicly available draft, which “reflects what was to become the standard at the time of issue”, according to [open-std.org](http://www.open-std.org/jtc1/sc22/wg14/www/standards.html). The official ISO page is https://www.iso.org/standard/29237.html.
5. Jean-Michel Muller, Nicolas Brisebarre, Florent de Dinechin, Claude-Pierre Jeannerod, Vincent Lefèvre, Guillaume Melquiond, Nathalie Revol, Damien Stehlé, and Serge Torres. _Handbook of Floating-Point Arithmetic_. Boston: Birkhäuser, 2010. [doi:10.1007/978-0-8176-4705-6](https://doi.org/10.1007/978-0-8176-4705-6), e-ISBN: 978-0-8176-4705-6.

--------------

†Of course, backward compatibility rules here. I'm not actually proposing in the middle of a different PR to rip out support for AIX's `xlc128` compiler that's already been around for awhile. The ship has sailed, the horse is out of the etc.


[`long double`]: https://numpy.org/doc/stable/user/basics.types.html#extended-precision
[AIX 7.1 documentation]: https://www.ibm.com/docs/en/aix/7.1?topic=sepl-128-bit-long-double-floating-point-data-type

Reply:
> This suggests that smallest_normal_dd is the same as the smallest, positive normal in a double.

True, that is what it was, but then `finfo` lies about the precision, if you say that every number larger than `smallest_normal` should have full (in this case at least) the number of mantissa bits advertised).  That would be 106bits (including the implicit one).

We have those two main choices: just go with double precision (the old value) or go with the smallest number that has a 106bit (implicit) mantissa (and above which all numbers do!).  Which actually aligns with what you said, and is `np.exp2(-969.0)` (although I am only 95% certain).

Double-double is not IEEE compliant, so we have to make a choice.  It would be nice to just point to someone else's choice here, but I don't know it.  I am perfectly fine with us making the choice here, if we are pretty confident about one.

Reply:
> We have those two main choices: just go with double precision (the old value) or go with the smallest number that has a 106bit (implicit) mantissa (and above which all numbers do!)

Upon further reflection the correct answer is the latter. I can explain more later but for now the basic idea is that the distinction between subnormal and normal is all about guaranteed minimum precision. Subnormal don't maintain the main axiom of FP in numerical analysis that _relative_ errors are bounded by half a machine epsilon. Normals do maintain that invariant. So setting the dividing line between them based on precision is the only useful way to go.

Reply:
I am confused. We had agreed to merge this PR with undefined values for long double. But now we are opening the discussion up again?

Reply:
No, I am happy to follow up here, we can move into an issue.